### PR TITLE
[DP-171] work on predict microservice

### LINKED ISSUE #804: Predict api production cicd pipeline failure
**What were you trying to achieve?**
This is happening due to a typo in the production workflow file
image

**What are the expected results?**

**What are the received results?**

**What are the steps to reproduce the issue?**

**In what environment did you encounter the issue?**

**Additional context**
Any other information you would like to share?

### PR DESCRIPTION
**Pull request by danielogen**
# work on predict microservice

**_WHAT DOES THIS PR DO?_**
- Change directory structure of predict microservice (move jobs to one directory)
- Refactor predict training job
- Add mlops configuration using mlflow
- Set automatic model registering, experiment tracking, logging parameters for each model run
- Automatic model backup and update
- Closes #804 

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
- [DP-171]
- [PLAT-1072](https://airqoteam.atlassian.net/browse/PLAT-1072)

**_WHAT IS THE LINK TO THE PR BRANCH_**
- https://github.com/airqo-platform/AirQo-api/tree/predict-train-refactor

**_HOW DO I TEST OUT THIS PR?_**

1. check out this branch and `cd` to `src/predict/jobs/train/`
2. using python virtual environment, run `pip install -r requirements.txt`. Make sure you have the `.env` file.
3. Run `python main.py` and wait for the process to finish. *NB: set the environment to `staging` to avoid overwriting models in production*.
4. Check the result on mlflow server: `http://23.251.144.212:5000`. You can also check the `model.pkl` and `all_channels.pkl` uploaded in google storage bucket: `airqo_prediction_bucket_stage`.

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 
- NONE

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
- NONE

**_ARE THERE ANY RELATED PRs?_**
- NONE

[DP-171]: https://airqoteam.atlassian.net/browse/DP-171?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/predict/jobs/train/Dockerfile [hunk 1]
```diff
@@ -7,9 +7,17 @@ COPY . /app/
 WORKDIR /app/
```
Comment (Baalmart) [inline src/predict/jobs/train/Dockerfile:7]: Just noticing this new directory for predict-job......care to share some background thoughts on this? @danielogen

**Review by Baalmart — COMMENTED** [review]
Thanks @danielogen :
Since there are many changed files, could you review the following alternatives: 
1. Add your PR review comments which will act guiding notes for the reviewers. Similar to how people comment in their code
2. Organise a short call for you to onboard us -- less desirable option.

**Review by sserurich — APPROVED** [review]
This LGTM, may be for the benefit of other reviewers, @danielogen , can consider updating .env in the platform/predict on google drive.
image

**Comment by danielogen** [comment]
> This LGTM, may be for the benefit of other reviewers, @danielogen , can consider updating .env in the platform/predict on google drive. image

Am on it. Thank you!

**Comment by uman95** [comment]
LGTM. 
Ran successfully.
Screenshot from 2022-04-04 11-34-59

**Review by uman95 — APPROVED** [review]
LGTM

File: k8s/predict/stage-airqo-train-job.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+apiVersion: batch/v1beta1
+kind: CronJob
+metadata:
+  name: stage-airqo-train-job
+  namespace: staging
+spec:
+  successfulJobsHistoryLimit: 1
+  failedJobsHistoryLimit: 1
+  schedule: "0 5 * * 1"
```
Comment (123MwanjeMike) [inline k8s/predict/stage-airqo-train-job.yaml:9]: Run the cron job weekly on Monday at 5am

**Review by 123MwanjeMike — COMMENTED** [review]
Please note that the GitGaurdian secrets discovered have already been changed.

**Review by NoahNsimbe — APPROVED** [review]
LGTM

**Pull request merged**